### ISSUE TITLE
How to use pack_padded_sequence with the existing sentiment analysis code?

### ISSUE THREAD

**Issue by jibinmathew691993** [bentrevett/pytorch-sentiment-analysis#5]
can someone point out the code to add pack_padded_sequence and it's unpacking?

**Comment by jibinmathew691993** [comment]
solved it, it can from float tensors instead of int

**Comment by jibinmathew691993** [comment]
how do pack_padded_sequnce work for lstms with dropout ? do they have to be unpacked before applying linear layers?

**Comment by bentrevett** [comment]
They should work the exact same way. 

You will have to unpack them before passing them to any layers after the LSTM.

**Comment by jibinmathew691993** [comment]
I tried without unpacking but it seemed to work fine, is there anything fundamentally wrong with that approach?

**Comment by bentrevett** [comment]
So, taking the model from tutorial 2 as an example, if we add packed padded sequences, it should be something like:

```python
class RNN(nn.Module):
    def __init__(self, vocab_size, embedding_dim, hidden_dim, output_dim, n_layers, bidirectional, dropout):
        super().__init__()
        
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.rnn = nn.LSTM(embedding_dim, hidden_dim, num_layers=n_layers, bidirectional=bidirectional, dropout=dropout)
        self.fc = nn.Linear(hidden_dim*2, output_dim)
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, text, text_lengths):
                
        embedded = self.dropout(self.embedding(text))
                
        packed_embedded = nn.utils.rnn.pack_padded_sequence(embedded, text_lengths)
        
        output, (hidden, cell) = self.rnn(packed_embedded)
                
        hidden = self.dropout(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1))
            
        return self.fc(hidden.squeeze(0))
```

Here, we don't need to unpack `output` as we don't use it for the dropout and linear layers.  The `hidden` and `cell` tensors never need to be unpacked as they are just tensors and not packed sequences.

**Comment by ZenRay** [comment]
@jibinmathew69 may I ask a question? I want to know why don't use unpacked output to run full connection, otherwise use `hidden`. Is there a specific reason?

**Comment by jibinmathew69** [comment]
@ZenRay If you mean why use pack padded sequence instead of unpacked, You don't want to waste computations looking at padding values and also it doesn't make sense.

**Comment by ZenRay** [comment]
My confusion is here, [2 - Upgraded Sentiment Analysis.ipynb](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/2%20-%20Upgraded%20Sentiment%20Analysis.ipynb) 

```
        hidden = self.dropout(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim = 1))
                
        #hidden = [batch size, hid dim * num directions]
            
        return self.fc(hidden)
```

why use the `hidden` value to compute the final output?

**Comment by MatthewSwahn** [comment]
@ZenRay the hidden value is the "last" output value. As per the docs https://pytorch.org/docs/stable/nn.html

> Outputs: output, (h_n, c_n)
> output of shape (seq_len, batch, num_directions * hidden_size): tensor containing the output features (h_t) from the last layer of the LSTM, for each t. If a torch.nn.utils.rnn.PackedSequence has been given as the input, the output will also be a packed sequence.
> 
> For the unpacked case, the directions can be separated using output.view(seq_len, batch, num_directions, hidden_size), with forward and backward being direction 0 and 1 respectively. Similarly, the directions can be separated in the packed case.
> 
> h_n of shape (num_layers * num_directions, batch, hidden_size): **tensor containing the hidden state for t = seq_len.**
> 
> Like output, the layers can be separated using h_n.view(num_layers, num_directions, batch, hidden_size) and similarly for c_n.

**Comment by ZenRay** [comment]
@MatthewSwahn thx,

**Comment by Dongximing** [comment]
> So, taking the model from tutorial 2 as an example, if we add packed padded sequences, it should be something like:
> 
> ```python
> class RNN(nn.Module):
>     def __init__(self, vocab_size, embedding_dim, hidden_dim, output_dim, n_layers, bidirectional, dropout):
>         super().__init__()
>         
>         self.embedding = nn.Embedding(vocab_size, embedding_dim)
>         self.rnn = nn.LSTM(embedding_dim, hidden_dim, num_layers=n_layers, bidirectional=bidirectional, dropout=dropout)
>         self.fc = nn.Linear(hidden_dim*2, output_dim)
>         self.dropout = nn.Dropout(dropout)
>         
>     def forward(self, text, text_lengths):
>                 
>         embedded = self.dropout(self.embedding(text))
>                 
>         packed_embedded = nn.utils.rnn.pack_padded_sequence(embedded, text_lengths)
>         
>         output, (hidden, cell) = self.rnn(packed_embedded)
>                 
>         hidden = self.dropout(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1))
>             
>         return self.fc(hidden.squeeze(0))
> ```
> 
> Here, we don't need to unpack `output` as we don't use it for the dropout and linear layers. The `hidden` and `cell` tensors never need to be unpacked as they are just tensors and not packed sequences. However, if you were using `output` further in the model, I believe you would need to unpack it.

Hi, how can I use "output" layer as hidden(linear) layer input? since in your code using "hidden"  as hidden(linear) layer input
